keel-igor/src/main/kotlin/com/netflix/spinnaker/keel/igor/BuildLifecycleMonitor.kt

```diff
@@ -0,0 +1,197 @@
+    } catch (e: IllegalArgumentException) {
+      publishUnknownEvent(task)
+      endMonitoringOf(task)
+      null
+   }
+
+  private fun parseBuildData(task: MonitoredTask): BuildData =
+      objectMapper.convertValue(task.triggeringEvent.data)
+
+
+  private fun BuildMetadata.isComplete(): Boolean =
+    status == "SUCCESS" || status == "FAILED" || status == "ABORTED"
+
+  private fun publishRunningEvent(task: MonitoredTask) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = RUNNING,
+      link = chooseLink(task),
+      text = "Build running for version ${task.triggeringEvent.artifactVersion}"
+    ))
+  }
+
+  private fun publishSucceededEvent(task: MonitoredTask, buildMetadata: BuildMetadata) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = SUCCEEDED,
+      link = chooseLink(task),
+      text = "Build succeeded for version ${task.triggeringEvent.artifactVersion}",
+      timestamp = buildMetadata.getCompletedAtInstant()
+    ))
+  }
+
+  private fun publishFailedEvent(task: MonitoredTask, buildMetadata: BuildMetadata) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = FAILED,
+      link = chooseLink(task),
+      text = "Build failed for version ${task.triggeringEvent.artifactVersion}",
+      timestamp = buildMetadata.getCompletedAtInstant()
+    ))
+  }
+
+  private fun publishAbortedEvent(task: MonitoredTask, buildMetadata: BuildMetadata) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = ABORTED,
+      link = chooseLink(task),
+      text = "Build aborted for version ${task.triggeringEvent.artifactVersion}",
+      timestamp = buildMetadata.getCompletedAtInstant()
+    ))
+  }
+
+  private fun publishUnknownStatusEvent(task: MonitoredTask, status: String?) {
+    log.warn("Unknown status $status while monitoring ${task.triggeringEvent}")
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = UNKNOWN,
+      link = chooseLink(task),
+      text = "Build status unknown for version ${task.triggeringEvent.artifactVersion}"
+    ))
+  }
+
+  private fun publishUnknownEvent(task: MonitoredTask) {
+    log.warn("Unable to monitor build for ${task.triggeringEvent} because at least one required data field is missing")
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = UNKNOWN,
+      link = chooseLink(task),
+      text = "Build status unknown for version ${task.triggeringEvent.artifactVersion}"
+    ))
+  }
+
+  override fun publishExceptionEvent(task: MonitoredTask) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = UNKNOWN,
+      link = chooseLink(task),
+      text = "Failed to monitor build of version ${task.triggeringEvent.artifactVersion}" +
+        " because we could not get the status ${lifecycleConfig.numFailuresAllowed} times. Status unknown."
+    ))
+  }
+
+  fun chooseLink(task: MonitoredTask): String? {
+    val buildData = parseBuildData(task)
+    val app = runBlocking {
```

Comment:
This is going to block the event dispatch thread while it makes a call to Fiddy? Seems like we could decouple that to avoid blocking.

Reply:
Ya. How would you recommend decoupling this? making it a suspend function?

Reply:
discussed offline, we can optimize later